pkg/sql/catalog/table_elements.go

```diff
@@ -408,30 +433,153 @@ type Constraint interface {
 	// GetConstraintValidity returns the validity of this constraint.
 	GetConstraintValidity() descpb.ConstraintValidity
 
+	// IsEnforced returns true iff the constraint is enforced for all writes to
+	// the parent table.
+	IsEnforced() bool
+
 	// GetName returns the name of this constraint update mutation.
 	GetName() string
 
-	// NotNullColumnID returns the underlying not-null column ID, if there is one.
-	NotNullColumnID() descpb.ColumnID
+	// IsConstraintValidated returns true iff the constraint is enforced for
+	// all writes to the table data and has also been validated on the table's
+	// existing data prior to the addition of the constraint, if there was any.
+	IsConstraintValidated() bool
+
+	// IsConstraintUnvalidated returns true iff the constraint is enforced for
+	// all writes to the table data but which is explicitly NOT to be validated
+	// on any data in the table prior to the addition of the constraint.
+	IsConstraintUnvalidated() bool
+}
+
+// UniqueConstraint is an interface for a unique constraint.
+type UniqueConstraint interface {
+	Constraint
+
+	// IsValidReferencedUniqueConstraint returns whether the unique constraint can
+	// serve as a referenced unique constraint for a foreign key constraint.
+	IsValidReferencedUniqueConstraint(fk ForeignKeyConstraint) bool
+
+	// NumKeyColumns returns the number of columns in this unique constraint.
+	NumKeyColumns() int
+
+	// GetKeyColumnID returns the ID of the column in the unique constraint at
+	// ordinal `columnOrdinal`.
+	GetKeyColumnID(columnOrdinal int) descpb.ColumnID
```

Comment:
Just to make sure, all the column ordinals we used in the codebase, they are 0-indexing, right?

Reply:
Yes, just like all `Get\w+ID(\w+Ordinal int)` methods in this package.